src/plugins/explore/public/application/pages/traces/trace_details/public/services/service_map.tsx

```diff
@@ -0,0 +1,1059 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import React, { useMemo, useCallback, useEffect, useState } from 'react';
+import { i18n } from '@osd/i18n';
+import {
+  EuiPanel,
+  EuiPanelProps,
+  EuiComboBox,
+  EuiComboBoxOptionOption,
+  EuiButton,
+  EuiCheckbox,
+  EuiTitle,
+  EuiSpacer,
+  EuiLoadingSpinner,
+  EuiPopover,
+  EuiContextMenu,
+  EuiIcon,
+} from '@elastic/eui';
+import {
+  ReactFlow,
+  ReactFlowProvider,
+  Node,
+  Edge,
+  Controls,
+  Background,
+  useNodesState,
+  useEdgesState,
+  ConnectionMode,
+  Handle,
+  Position,
+  useReactFlow,
+  MarkerType,
+} from '@xyflow/react';
+import dagre from '@dagrejs/dagre';
+import '@xyflow/react/dist/style.css';
+import './service_map.scss';
+
+interface SpanHit {
+  spanId: string;
+  parentSpanId: string;
+  serviceName: string;
+  durationInNanos: number;
+  status?: {
+    code: number;
+    message: string;
+  };
+}
+
+interface ServiceNodeData {
+  label: string;
+  spanCount: number;
+  avgLatency: number;
+  maxLatency: number;
+  errorRate: number;
+  color: string;
+  maxValues?: {
+    maxRequestRate: number;
+    maxErrorRate: number;
+    maxDuration: number;
+  };
+  [key: string]: unknown;
+}
+
+// Custom node type definition
+type ServiceNodeType = Node<ServiceNodeData>;
+
+// Metric display options
+type MetricOption = 'duration' | 'errorRate' | 'requestRate';
+
+// Format latency values - always display in milliseconds
+const formatLatency = (nanos: number) => {
+  const ms = nanos / 1000000;
+  if (ms >= 1000) {
+    return `${(ms / 1000).toFixed(2)}s`;
+  } else if (ms >= 1) {
+    return `${ms.toFixed(1)}ms`;
+  } else {
+    return `${ms.toFixed(2)}ms`;
+  }
+};
+
+const ServiceNode = ({
+  data,
+  selectedMetrics,
+  showDetails,
+  onFocusService,
+  onToggleDetails,
+  openPopoverNodeId,
+  setOpenPopoverNodeId,
+  isSelected,
+}: {
+  data: ServiceNodeData;
+  selectedMetrics: MetricOption[];
+  showDetails: boolean;
+  onFocusService: (serviceName: string) => void;
+  onToggleDetails: () => void;
+  openPopoverNodeId: string | null;
+  setOpenPopoverNodeId: (id: string | null) => void;
+  isSelected?: boolean;
+}) => {
+  const serviceColor = data.color;
+
+  // Calculate intensity values based on max values
+  const requestRateIntensity = data.spanCount / (data.maxValues?.maxRequestRate || 100);
```

Comment:
Curious can the data.maxValues?.maxRequestRate ever be 0?

Reply:
It should never be undefined or 0 as the service map is only generated if we have valid spans making up the trace. This calculation is for making the colored bars of the service map to give a representation of it's weight in the trace.

These fallbacks are just reasonable numbers for the fields to ensure that the component does not crash as filter are applied and it is re-rendered.

<img width="205" height="182" alt="Screenshot 2025-07-28 at 10 45 21 AM" src="https://github.com/user-attachments/assets/daf6837a-fa2e-44f4-bc9f-e52ea7be5859" />


<img width="1120" height="376" alt="Screenshot 2025-07-28 at 10 44 45 AM" src="https://github.com/user-attachments/assets/fe89a82b-10ec-4937-b970-ea4e8ea0825f" />